## Hypothesis Testing

In [1]:
# Import Libraries:

import pandas as pd
import numpy as py
import tmdbsimple as tmdb

In [2]:
import json
with open('/Users/jnate/.secret/tmdb_api.json', 'r') as f:
    login = json.load(f)
#Display the keys of the loaded dict:
login.keys()

dict_keys(['api-key'])

In [3]:
# Load API key:

tmdb.API_KEY =  login['api-key']

In [4]:
# Load JSON files:

with open('Data\\tmdb_api_results_2019.json') as y1:
    jsonfile1 = json.load(y1)

with open('Data\\tmdb_api_results_2020.json') as y2:
    jsonfile2 = json.load(y2)

In [40]:
# Convert separate years to pandas DataFrames:

year2019 = pd.DataFrame(jsonfile1)

# Reduce the DataFrame down to 4 columns:
year2019 = year2019[['imdb_id', 'revenue', 'budget', 'certification']]

# Imputation:

year2019['certification'].replace({'' : 'NR', '' : 'NR'}, inplace = True);

year2019['certification'].fillna('NR', inplace =True)

year2019['revenue'].fillna(0.0, inplace =True)

year2019['budget'].fillna(round(year2019['budget'].mean(),2), inplace = True)

# Check:

year2019.head()

imdb_id      revenue        budget certification
0          0          0.0  1.914495e+06            NR
1  tt0385887   18377736.0  2.600000e+07             R
2  tt0437086  404852543.0  1.700000e+08         PG-13
3  tt0441881    2078370.0  2.393482e+07             R
4  tt0448115  365971656.0  8.000000e+07         PG-13

In [41]:
# Convert separate years to pandas DataFrames:

year2020 = pd.DataFrame(jsonfile2)

# Reduce the DataFrame down to 4 columns:
year2020 = year2020[['imdb_id', 'revenue', 'budget', 'certification']]

# Imputation:

year2020['certification'].replace({'' : 'NR', '' : 'NR'}, inplace = True);

year2020['certification'].fillna('NR', inplace =True)

year2020['revenue'].fillna(0.0, inplace =True)

year2020['budget'].fillna(round(year2020['budget'].mean(),2), inplace = True)

# Check:

year2020.head()

imdb_id     revenue      budget certification
0          0         0.0  1026934.58            NR
1  tt0062336         0.0        0.00            NR
2  tt0093119         0.0  7500000.00            NR
3  tt0805647  26900000.0        0.00            PG
4  tt0920462    106045.0        0.00             R

In [42]:
# Check for duplciates:

year2019.duplicated().sum()

0

In [43]:
# Check for duplciates:

year2020.duplicated().sum()

0

In [44]:
year2019.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4677 entries, 0 to 4676
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   imdb_id        4677 non-null   object 
 1   revenue        4677 non-null   float64
 2   budget         4677 non-null   float64
 3   certification  4677 non-null   object 
dtypes: float64(2), object(2)
memory usage: 146.3+ KB


In [45]:
year2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3718 entries, 0 to 3717
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   imdb_id        3718 non-null   object 
 1   revenue        3718 non-null   float64
 2   budget         3718 non-null   float64
 3   certification  3718 non-null   object 
dtypes: float64(2), object(2)
memory usage: 116.3+ KB


In [46]:
# Remvoe NaN values:

year2019.dropna(inplace = False)
year2020.dropna(inplace = False)

imdb_id     revenue      budget certification
0             0         0.0  1026934.58            NR
1     tt0062336         0.0        0.00            NR
2     tt0093119         0.0  7500000.00            NR
3     tt0805647  26900000.0        0.00            PG
4     tt0920462    106045.0        0.00             R
...         ...         ...         ...           ...
3713  tt9904004         0.0        0.00            NR
3714  tt9904802         0.0        0.00            NR
3715  tt9911196         0.0        0.00            NR
3716  tt9916190         0.0        0.00            NR
3717  tt9916362         0.0        0.00            NR

[3718 rows x 4 columns]

In [47]:
year2019.isna().sum()

imdb_id          0
revenue          0
budget           0
certification    0
dtype: int64

In [48]:
# Combined DataFrame:

# Combine / Concatinate separate DataFrames: 

combined = [year2019, year2020]

df = pd.concat(combined)

df.head()

imdb_id      revenue        budget certification
0          0          0.0  1.914495e+06            NR
1  tt0385887   18377736.0  2.600000e+07             R
2  tt0437086  404852543.0  1.700000e+08         PG-13
3  tt0441881    2078370.0  2.393482e+07             R
4  tt0448115  365971656.0  8.000000e+07         PG-13

In [49]:
# Removing movies with 0 revenue: 

no_revenue = df.loc[df['revenue']==0]

# df[~no_revenue]

In [50]:
# Excluding titles with a budget of 0. 

budget = df.loc[df['budget'].copy() != 0]

df = budget

# Excluding titles with 0 revenue:

revenue = df.loc[df['revenue'].copy() != 0]

df = revenue

df.head()

imdb_id      revenue       budget certification
1  tt0385887   18377736.0   26000000.0             R
2  tt0437086  404852543.0  170000000.0         PG-13
3  tt0441881    2078370.0   23934823.0             R
4  tt0448115  365971656.0   80000000.0         PG-13
5  tt0783640    3364426.0   20000000.0             R

In [51]:
# Exporting the final Data Frame as csv:

df.to_csv("Data/year2019_2020.csv.gz",compression='gzip',index=False)

In [52]:
df['certification'].value_counts()

NR       118
R         84
PG-13     54
PG        35
G          4
NC-17      1
Name: certification, dtype: int64

### Question: Does the MPAA rating of a movie (G/PG/PG-13/R) affect how much revenue the movie generates?

***1. State your Null Hypothesis and Alternative Hypothesis:***

$H_0$: MPAA rating does not affect a movie's revenue.

$H_A$: MPAA does affect a movie's revenue.

***2. Select the correct test according to the data type and number of samples.***

    Type of comparison: more than 2 samples

    Data type: Numeric

    Hypothesis Test: ANOVA and/or Tukey

### Assumptions of the selected test:

    No significant outliers
    Equal variance
    Normality

In [53]:
# Create dictionary for the various regions:

ratings = {}

# Loop through column to get the unique designations:

for i in df['certification'].unique():
    # Get series for group and rename:
    data = df.loc[df['certification']==i, 'revenue'].copy()
    
    # Append dictionary:
    ratings[i] = data
ratings.keys()

dict_keys(['R', 'PG-13', 'NR', 'NC-17', 'PG', 'G'])

In [54]:
# # Removing certifications with n < 3 observations 

# del ratings['NC-17']

# ratings.keys()

In [55]:
import numpy as np
from scipy import stats

# Check for significant outliers:

for rating_name, revenue_df in ratings.items():
    # Identify outliers:
    outliers = np.abs(stats.zscore(revenue_df))>3
    
    # Outliers per region:
    print(f"There were {outliers.sum()} outliers in the {rating_name} group.")
    
    # Remove the outliers from dat and overwrite teh region dictionary:
    revenue_df = revenue_df[~outliers]
    
    ratings[rating_name] = revenue_df

There were 1 outliers in the R group.
There were 1 outliers in the PG-13 group.
There were 2 outliers in the NR group.
There were 0 outliers in the NC-17 group.
There were 2 outliers in the PG group.
There were 0 outliers in the G group.


In [56]:
# Certification n value:

for certification, n_value in ratings.items():
    print(certification, ":", len(n_value))

R : 83
PG-13 : 53
NR : 116
NC-17 : 1
PG : 33
G : 4


In [57]:
# Normality Test:

# Normal test for each region to confirm that there are more than 20 samples in each group.
norm_results = [['certification', 'n', 'p', 'significance?']]

# Loop through each certification in the dictionary:

for ratings_name, temp_df in ratings.items():
    n = len(temp_df)
    if n < 3:
        n = 'n/a'
    else:
        
        # Calculate normal test reuslts:
        result = stats.shapiro(temp_df)
        p = result.pvalue
        n = len(temp_df)
    
    # Appred the normal results as list:
    norm_results.append([ratings_name, n, p, p<.05])
nrdf = pd.DataFrame(norm_results[1:], columns = norm_results[0])
nrdf

certification    n             p  significance?
0             R   83  2.167157e-12           True
1         PG-13   53  1.397880e-09           True
2            NR  116  1.363012e-15           True
3         NC-17  n/a  1.363012e-15           True
4            PG   33  4.865038e-07           True
5             G    4  4.169193e-03           True

#### Interpretation:

p less than alpha for all categories - we do not have normal distributions. We must move on to selecting a non-parametric equivalent test. 

In [58]:
# Kruskal-Wallis Test for the various movie categories:

kw_result = stats.kruskal(*ratings.values())
kw_result

KruskalResult(statistic=73.92697971857233, pvalue=1.5577054948795198e-14)

In [59]:
# If-else statement to interpret the p-value:

if kw_result.pvalue < 0.05:
    print('p is less than alpha, reject null hypothesis')
    
else:
    print('p is greater than alpah, the null hypothesis is validated')

p is less than alpha, reject null hypothesis


### ***Interpretation:***

p = 1.558 x 10 e-14
alpha = 0.05

p is significantly smaller than alpha meaning that we can reject the null hypothesis and support the alternative hypothesis; MPAA rating does affect a movie's revenue.

### Question: Do movies released in 2020 earn less revenue than movies released in 2019?

### How do the years compare for movie ratings?

***1. State your Null Hypothesis and Alternative Hypothesis:***

$H_0$: Movies released in 2020 do not earn less than movies released in 2019.

$H_A$: Movies released in 2020 make significantly less than movies released in 2019.

***2. Select the correct test according to the data type and number of samples.***

    Type of comparison: 2 samples

    Data type: Numeric

    Hypothesis Test: 2 Sample T-Test

### Assumptions of the selected test:

    No significant outliers
    Normality
    Equal Variance

In [66]:
# Define the two groups:


revenue_2019 = year2019.loc[year2019['revenue']!=0]
revenue_2020 = year2020.loc[year2020['revenue']!=0]

budget_2019 = year2019.loc[year2019['budget']!=0]
budget_20202 = year2020.loc[year2020['budget']!=0]

y2019 = year2019[revenue_2019, budget_2019]
y2020 = year2020.loc[revenue_2020, budget_2020]

TypeError: '(        imdb_id      revenue       budget certification
1     tt0385887   18377736.0   26000000.0             R
2     tt0437086  404852543.0  170000000.0         PG-13
3     tt0441881    2078370.0   23934823.0             R
4     tt0448115  365971656.0   80000000.0         PG-13
5     tt0783640    3364426.0   20000000.0             R
...         ...          ...          ...           ...
4533  tt9786512    2197075.0          0.0            NR
4543  tt9795368    1964806.0          0.0             R
4583  tt9826484      15712.0          0.0             R
4600  tt9845110     208723.0          0.0            NR
4601  tt9845398     103047.0          0.0            NR

[384 rows x 4 columns],         imdb_id      revenue        budget certification
0             0          0.0  1.914495e+06            NR
1     tt0385887   18377736.0  2.600000e+07             R
2     tt0437086  404852543.0  1.700000e+08         PG-13
3     tt0441881    2078370.0  2.393482e+07             R
4     tt0448115  365971656.0  8.000000e+07         PG-13
...         ...          ...           ...           ...
4626  tt9866202          0.0  1.100000e+06            NR
4628  tt9869668          0.0  5.000000e+03            NR
4634  tt9875852          0.0  2.310000e+06            NR
4640  tt9881572          0.0  1.000000e+04            NR
4668  tt9903890          0.0  3.500000e+03            NR

[587 rows x 4 columns], slice(None, None, None))' is an invalid key

### Test for significant outliers

In [21]:
# Check for outliers using Z-scores:

y19_outliers = np.abs(stats.zscore(y19))>3
y20_outliers = np.abs(stats.zscore(y20))>3

y19_outliers.value_counts()

revenue
False      4645
True         32
dtype: int64

In [22]:
y20_outliers.value_counts()

revenue
False      3701
True         17
dtype: int64

In [28]:
# Removal of outliers:

y19_final = y19[-y19_outliers]
y20_final = y20[-y20_outliers]

y19_final

revenue
0            0.0
1     18377736.0
2            NaN
3      2078370.0
4            NaN
...          ...
4672         0.0
4673         0.0
4674         0.0
4675         0.0
4676         0.0

[4677 rows x 1 columns]

### Normality Test

In [24]:
# Normality test for 2019:

NT2019 = stats.normaltest(y19_final)
NT2019

NormaltestResult(statistic=array([nan]), pvalue=array([nan]))

In [25]:
# Normality test for 2019:

NT2020 = stats.normaltest(y20_final)
NT2020

NormaltestResult(statistic=array([nan]), pvalue=array([nan]))

### Question: Do some movie genres earn more revenue than others?

***1. State your Null Hypothesis and Alternative Hypothesis:***

$H_0$: genre classification does not play a significant role in move revenue.

$H_A$: genre classification has a significant influence on revenue.

***2. Select the correct test according to the data type and number of samples.***

    Type of comparison: more than 2 samples

    Data type: Numeric

    Hypothesis Test: ANOVA and/or Tukey

### Assumptions of the selected test:

    No significant outliers
    Normality
    Equal Variance

In [26]:
# Create dictionary for the various regions:

ratings = {}

# Loop through column to get the unique designations:

for i in df['certification'].unique():
    # Get series for group and rename:
    data = df.loc[df['certification']==i, 'revenue'].copy()
    
    # Append dictionary:
    ratings[i] = data
ratings.keys()

dict_keys(['R', 'PG-13', 'NR', 'NC-17', 'PG', 'G'])

In [27]:
# Check for significant outliers:

for rating_name, revenue_df in ratings.items():
    # Identify outliers:
    outliers = np.abs(stats.zscore(revenue_df))>3
    
    # Outliers per region:
    print(f"There were {outliers.sum()} outliers in the {rating_name} group.")
    
    # Remove the outliers from dat and overwrite teh region dictionary:
    revenue_df = revenue_df[~outliers]
    
    ratings[rating_name] = revenue_df

There were 1 outliers in the R group.
There were 1 outliers in the PG-13 group.
There were 2 outliers in the NR group.
There were 0 outliers in the NC-17 group.
There were 2 outliers in the PG group.
There were 0 outliers in the G group.
